baremetal/metal3data_manager.go

```diff
@@ -1373,6 +1377,32 @@ func getBMHMacByName(name string, bmh *bmov1alpha1.BareMetalHost) (string, error
 	return "", fmt.Errorf("nic name not found %v", name)
 }
 
+func getValueFromAnnotation(object string, annotation string,
+	m3m *infrav1.Metal3Machine, machine *clusterv1.Machine, bmh *bmov1alpha1.BareMetalHost) (string, error) {
+	switch strings.ToLower(object) {
+	case m3machine:
+		return m3m.Annotations[annotation], nil
+	case capimachine:
+		return machine.Annotations[annotation], nil
+	case host:
+		return bmh.Annotations[annotation], nil
+	default:
+		return "", errors.New("Unknown object type")
+	}
+}
+
+func getValueFromAnnotationStrict(object string, annotation string,
```

Comment:
I think we should just conclude that this is how we should handle annotations and get rid of `getValueFromAnnotation` (used when rendering metadata). Different behaviors for metadata and network data will just be confusing.

Reply:
/cc @kashifest

Reply:
During developing I found out that `getValueFromAnnotation` yields empty strings for non-existing annotations. (It is not clear to me whether Kubernetes itself distinguishes between empty or non-existing annotations.)

An empty MAC address is surely an error condition, and hence I decided to make that strict. (You could even argue to check the annotation further e.g. via a regular expression to allow early detection of errors.)

But for meta data it is perfectly valid to have empty values – we do just not know how the data is applied.

Reply:
Thanks for the explanation, that makes sense!
Could you add doc strings to these functions with some description that makes it clear what the difference is?

Reply:
This discussion made be think about that matter again and I have now added an explicit regexp check for the resulting MAC address – regardless of its source. That appears far more conclusive then the rather erratic and not self-explanatory distinction between potentially empty and non-empty annotations.

That means a subtle change of the overall behaviour: non-well-formed MAC addresses will not raise issues later after provisioning, but already before.